code/modules/ai/ai_behaviors/xeno/xeno.dm

```diff
@@ -0,0 +1,131 @@
+			xeno_action.use_ability(atom_to_walk_to)
+		else
+			action.action_activate()
+	return ..()
+
+/datum/ai_behavior/xeno/look_for_new_state()
+	switch(current_action)
+		if(ESCORTING_ATOM, MOVING_TO_NODE)
+			var/atom/next_target = get_nearest_target(escorted_atom, target_distance, TARGET_ALL, null, mob_parent.get_xeno_hivenumber())
+			if(!next_target)
+				return
+			change_action(MOVING_TO_ATOM, next_target)
+		if(MOVING_TO_ATOM)
+			if(escorted_atom && get_dist(escorted_atom, mob_parent) > target_distance)
+				change_action(ESCORTING_ATOM, escorted_atom)
+				return
+			var/atom/next_target = get_nearest_target(escorted_atom, target_distance, TARGET_ALL, null, mob_parent.get_xeno_hivenumber())
+			if(!next_target)//We didn't find a target
+				cleanup_current_action()
+				late_initialize()
+				return
+			if(next_target == atom_to_walk_to)//We didn't find a better target
+				return
+			change_action(null, next_target)//We found a better target, change course!
+
+///Signal handler to try to attack our target
+/datum/ai_behavior/xeno/proc/attack_target(datum/soure, atom/attacked)
+	SIGNAL_HANDLER
+	if(world.time < mob_parent.next_move)
+		return
+	if(!attacked)
+		attacked = atom_to_walk_to
+	if(get_dist(attacked, mob_parent) > 1)
+		return
+	mob_parent.face_atom(attacked)
+	if(isobj(attacked) && !(attacked.resistance_flags & XENO_DAMAGEABLE))
+		stack_trace("A xenomorph tried to attack a [attacked.name] that isn't considered XENO_DAMAGABLE according to resistance flags.")
+		return
+	var/mob/living/carbon/xenomorph/xeno = mob_parent
+	attacked.attack_alien(xeno, xeno.xeno_caste.melee_damage * xeno.xeno_melee_damage_modifier)
+	xeno.changeNext_move(xeno.xeno_caste.attack_delay)
+
+/* OLD CODE, REMOVED IN ANOTHER PR DISREGARD
+/datum/ai_behavior/carbon/xeno/deal_with_obstacle()
+	if(world.time < mob_parent.next_move)
+		return
+
+	var/list/things_nearby = range(mob_parent, 1) //Rather than doing multiple range() checks we can just archive it here for just this deal_with_obstacle
+	for(var/obj/structure/obstacle in things_nearby)
+		if(obstacle.resistance_flags & XENO_DAMAGEABLE)
+			var/mob/living/carbon/xenomorph/xeno = mob_parent
+			INVOKE_ASYNC(obstacle, /atom.proc/attack_alien, xeno)
+			mob_parent.face_atom(obstacle)
+			xeno.changeNext_move(xeno.xeno_caste.attack_delay)
+			return
+
+	//Cheat mode: insta open airlocks
+	for(var/obj/machinery/door/airlock/lock in things_nearby)
+		if(!lock.density) //Airlock is already open no need to force it open again
+			continue
+		if(lock.operating) //Airlock already doing something
+			continue
+		if(lock.welded) //It's welded, can't force that open
+			continue
+		lock.open(TRUE)
+		return //Don't try going on window frames after opening up airlocks dammit
+
+	//Teleport onto those window frames, we also can't attempt to attack said window frames so this isn't in the obstacles loop
+	for(var/obj/structure/window_frame/frame in things_nearby)
+		mob_parent.loc = frame.loc
+		return
+
+*/
```

Comment:
/* OLD CODE, REMOVED IN ANOTHER PR DISREGARD
```suggestion
*/
```
/datum/ai_behavior/carbon/xeno/deal_with_obstacle()
	if(world.time < mob_parent.next_move)
		return
	var/list/things_nearby = range(mob_parent, 1) //Rather than doing multiple range() checks we can just archive it here for just this deal_with_obstacle
	for(var/obj/structure/obstacle in things_nearby)
		if(obstacle.resistance_flags & XENO_DAMAGEABLE)
			var/mob/living/carbon/xenomorph/xeno = mob_parent
			INVOKE_ASYNC(obstacle, /atom.proc/attack_alien, xeno)
			mob_parent.face_atom(obstacle)
			xeno.changeNext_move(xeno.xeno_caste.attack_delay)
			return
	//Cheat mode: insta open airlocks
	for(var/obj/machinery/door/airlock/lock in things_nearby)
		if(!lock.density) //Airlock is already open no need to force it open again
			continue
		if(lock.operating) //Airlock already doing something
			continue
		if(lock.welded) //It's welded, can't force that open
			continue
		lock.open(TRUE)
		return //Don't try going on window frames after opening up airlocks dammit
	//Teleport onto those window frames, we also can't attempt to attack said window frames so this isn't in the obstacles loop
	for(var/obj/structure/window_frame/frame in things_nearby)
		mob_parent.loc = frame.loc
		return
*/